### ISSUE TITLE
Script doesn't allow students to proceed after error

### ISSUE THREAD

**Issue by sthoward0914** [MicrosoftLearning/DP-300T00-Administering-Relational-Databases-on-Azure#67]
# Module: 07, exercise 2
#### Step: 03

Description of issue

The script checks for the existence of the credential, and only executes the CREATE CREDENTIAL if the credential doesn't exist. However; students sometimes execute the script before they realize that 'sas_token' should be replaced with the actual sas token. When they do this, if they replace the sas_token in the single quotes with the actual value, the logic of the script is such that the CREATE CREDENTIAL statement will never execute, so they are not able to proceed on to step 5 successfully.

In order to allow the students to re-run the code snippet and correct an issue they may have created by not using the correct SAS token first, the script should check to see if the credential exists, then drop it if it does and re-create it. My suggestion is:

IF EXISTS 

(SELECT * FROM sys.credentials 

WHERE name = 'https://dp300storage.blob.core.windows.net/backups') 

BEGIN

DROP CREDENTIAL [https://dp300storage.blob.core.windows.net/backups]

END

GO

CREATE CREDENTIAL [https://dp300storage.blob.core.windows.net/backups]

WITH IDENTITY = 'SHARED ACCESS SIGNATURE',

SECRET = 'sas_token'

GO 

Repro steps:

1. Run step 3, but use the wrong SAS key, or just don't replace sas_token with a sas token. 
2. Run step 5. It will now fail.
3. Realize you should have replaced sas_token in step 3 with an actual sas token. Replace it and run step 3 again. 
4. Re-run step 5. It will still fail because re-running step 3 with the correct token did not change the credential.
5. Try replacing the script in step 3 with the suggested code and repeat what I've put in the 4 items above. this time, when you re-run step 5 in the lab instructions, it will succeed because it drops the credential that was mistakenly created and re-creates iwt with the right credential.

**Comment by kalendelaney** [comment]
It would be helpful if the repro was complete, and actually said what step 3 and step actually are.

**Comment by sthoward0914** [comment]
Kalen, I'm unsure how to explain it better or more complete unless I just demo it to you. The bottom line is that if a student doesn't realize that the literal value "sas key" should be replaced with the SAS key from their actual storage account, they create an incorrect credential, and the backup to URL will fail. If they try to fix this using the scrip as it exists, re-running the script will not fix the issue to allow them to be successful in backing up to URL. The reason is because the script as it exists creates a credential if no credential by that name exists, but does nothing if a credential by that name already exists, even though that existing credential may not be correct.

<script as it exists>
IF NOT EXISTS 

(SELECT * FROM sys.credentials 

WHERE name = 'https://dp300storage.blob.core.windows.net/backups') 

BEGIN

CREATE CREDENTIAL [https://dp300storage.blob.core.windows.net/backups]

WITH IDENTITY = 'SHARED ACCESS SIGNATURE',

SECRET = 'sas_token'

END;

GO 
</script as it exists>

The difference between what is in the lab currently, and what I suggested as a fix is that, instead of doing "IF NOT EXISTS" as the current script does, my suggestion checks "IF EXISTS" and drops the existing credential if it is there. Changing the evaluation from 

<pseudocode>
IF NOT EXISTS 
(create the credential) 
</pseudocode>
to 
<pseudocode>
IF EXISTS 
(drop the credential)
THEN
Create the credential with the new information
</pseudocode>

allows the student to fix the credential. Once the credential is fixed, they can then be successful at backing up to URL using that credential. If the script in Lab 7, Task 2, Step 3 is replace in its entirety with the script I gave in the repro above, a student first entering an invalid credential, then wanting to fix that credential is no longer a problem an instructor needs to intervene on. Then steps I gave below that are just something students do sometimes. If they are followed, you will run into the same failure students run into on my repro step 2. Re-running the steps that currently exist in the code block in Lab 7, Task 2, step 3 will not fix the problem. But if that script is replaced with the one I provided, then if students need to do that step again to fix the credential, it will be successful.

I'd be happy to demo it for anybody if it is unclear from this explanation.

**Comment by kalendelaney** [comment]
Ok. I don't know if I can explain what was confusing any better, but the repo should have the exact steps. You just said run step 3, without saying what step 3 was. And the same for step 5. I had to go look up the steps in the lab instructions to see what you were referring to.  A problem report should have a complete repo that allows someone to completely understand the problem, without having to look at other sources.  But after looking at the full lab description, I see what you're saying. 

I approve of the idea of dropping the credential if it already exists, as opposed to just checking for existence before creation.  The DROP has been added to the lab step. Thanks
